controllers/invitation_cleanup_controller.go

```diff
@@ -0,0 +1,83 @@
+package controllers
+
+import (
+	"context"
+	"time"
+
+	apimeta "k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	userv1 "github.com/appuio/control-api/apis/user/v1"
+)
+
+// InvitationCleanupReconciler reconciles invitations, deleting them if appropriate
+type InvitationCleanupReconciler struct {
+	client.Client
+
+	Recorder record.EventRecorder
+	Scheme   *runtime.Scheme
+
+	RedeemedInvitationTTL time.Duration
+}
+
+//+kubebuilder:rbac:groups="rbac.appuio.io",resources=invitations,verbs=get;list;watch
+//+kubebuilder:rbac:groups="user.appuio.io",resources=invitations,verbs=get;list;watch
+//+kubebuilder:rbac:groups="rbac.appuio.io",resources=invitations/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups="user.appuio.io",resources=invitations/status,verbs=get;update;patch
+
+// Reconcile reacts on invitations and removes them if required
+func (r *InvitationCleanupReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+	log.V(4).WithValues("request", req).Info("Reconciling")
+
+	log.V(4).Info("Getting the Invitation...")
+	inv := userv1.Invitation{}
+	if err := r.Get(ctx, req.NamespacedName, &inv); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if !inv.ObjectMeta.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	}
+
+	if inv.Status.Token == "" {
+		// Invitation is not yet valid
+		return ctrl.Result{}, nil
+	}
+	now := metav1.NewTime(time.Now())
+
+	if inv.IsRedeemed() {
+		cond := apimeta.FindStatusCondition(inv.Status.Conditions, userv1.ConditionRedeemed)
+		ttlExpirationTime := cond.LastTransitionTime.Add(r.RedeemedInvitationTTL)
+		if ttlExpirationTime.Before(now.Time) {
+			log.V(4).Info("Redeemed Invitation TTL expired - deleting", "ttlExpirationTime", ttlExpirationTime)
```

Comment:
Does `V(4)` ever log anything? I had a faint memory of it only going to `V(2)` but i might be wrong.

Reply:
`V(4)` is used everywhere else as well. I'll check.

Reply:
It's not printed by default even on debug level - but it can in theory be set to log at level 4 via command line.
Should we change it everywhere?

Reply:
I seem to have copied without think all the time 😅

Yes i think we should change it but in a follow up PR.